loki/bulk/scheduler.py

```diff
@@ -102,97 +154,84 @@ def __init__(self, paths, config=None, seed_routines=None, preprocess=False,
             'includes': self.build_args['includes'],
             'defines': self.build_args['defines'],
-            'parser_classes': RegexParserClass.AllClasses,
+            'parser_classes': RegexParserClass.ProgramUnitClass,
             'frontend': REGEX
         }
 
         # Create a list of initial files to scan with the fast REGEX frontend
         path_list = [path.glob(f'**/*{ext}') for path in self.paths for ext in self.source_suffixes]
         path_list = list(set(flatten(path_list)))  # Filter duplicates and flatten
 
-        # Perform the full initial scan of the search space with the REGEX frontend
-        obj_list = [Sourcefile.from_file(filename=f, **frontend_args) for f in path_list]
-
-        debug(f'Total number of lines parsed: {sum(obj.source.lines[1] for obj in obj_list)}')
-
-        # Create a map of all potential target objs for fast lookup later
-        self.obj_map = CaseInsensitiveDict(
-            (f'#{r.name}', obj) for obj in obj_list for r in obj.subroutines
-        )
-        self.obj_map.update(
-            (f'{module.name}#{r.name}', obj)
-            for obj in obj_list for module in obj.modules
-            for r in module.subroutines + module.typedefs + module.variables
-        )
-        self.obj_map.update(
-            (f'{module.name}#{r.spec.name}', obj)
-            for obj in obj_list for module in obj.modules
-            for r in module.interfaces if r.spec
-        )
-
-    @property
-    def routines(self):
-        return as_tuple(item.routine for item in self.item_graph.nodes if item.routine is not None)
+        # Instantiate FileItem instances for all files in the search path
+        for path in path_list:
+            self.item_factory.get_or_create_file_item_from_path(path, self.config, frontend_args)
+
+        # Instantiate the basic list of items for files and top-level program units
+        #  in each file, i.e., modules and subroutines
+        #  Note that we do this separate from the FileItem instantiation above to enable discovery
+        #  also for FileItems that have been created as part of a transformation
+        file_items = [
+            file_item
+            for file_item in self.item_factory.item_cache.values()
+            if isinstance(file_item, FileItem)
+        ]
+        for file_item in file_items:
+            definition_items = {
+                item.name: item
+                for item in file_item.create_definition_items(item_factory=self.item_factory, config=self.config)
+            }
+            self.item_factory.item_cache.update(definition_items)
 
     @property
-    def typedefs(self):
-        return as_tuple(flatten(module.typedefs for obj in self.obj_map.values() for module in obj.modules))
+    def sgraph(self):
+        """
+        Create and return the :any:`SGraph` constructed from the :attr:`seeds` of the Scheduler.
+        """
+        return SGraph.from_seed(self.seeds, self.item_factory, self.config)
```

Comment:
As discussed offline, this automatic re-generation of the SGraph is very expensive at the moment, even though it does seem to correctly use the caching mechanism provided by the `ItemFactory`. My suspicion is the use of `Item.dependencies` triggering a complete re-parse of each edge/node of the base graph via the REGEX frontend, as this always seems to trigger `Item.concretize_dependencies`.

In this current stage this effect makes prohibitively expensive (enrichment triggers about 6 rebuilds for me at around 40s each in ec-physics), which can be mitigated by not caching the sgraph. However, I wonder if there is a way to not "concretize" things that are already concrete, and if this would alleviate the problem somewhat?

Reply:
I think you might be right!
The intention is to call make_complete to ensure the relevant parser classes have been parsed already - which should be a no-op after the initial discovery. For that reason, there is this bail-out here:

https://github.com/ecmwf-ifs/loki/blob/b1018476bb6408ee3b96adb5880968b41af0ebca/loki/program_unit.py#L290

Or at least, that's what it is intended to be. We call make_complete only with the `Item._depends_class`, which ideally will be a subset of what has been parsed already. But that's not what we are checking for... If my half-asleep brain is not completely wrong, this line should instead read as:
```python
            if self._parser_classes & parser_classes:
```

Would you mind trialling this?